File: pkg/R/whitening.R
## ID: whitening.R, last updated 2021-04-04, F.Osorio

whitening <- function(x, Scatter = NULL)
{ # Cholesky whitening transformation
  if (is.data.frame(x))
    x <- as.matrix(x)
  else if (!is.matrix(x))
    stop("'x' must be a matrix or a data frame")
  if (!all(is.finite(x)))
    stop("'x' must contain finite values only")
  storage.mode(x) <- "double"
  n <- nrow(x)
  p <- ncol(x)

  if (is.null(Scatter))
    Scatter <- cov.weighted(x, weights = rep(n / (n - 1), n))$cov

  if (!is.matrix(Scatter))
    stop("'Scatter' must be a matrix")
  if (nrow(Scatter) != ncol(Scatter))
    stop("'Scatter' must be a square matrix")
  if (p != ncol(Scatter))
    stop("'x' and 'Scatter' are not conformable")
  storage.mode(Scatter) <- "double"

  z <- .C("whitening_chol",
          x = x,
          n = as.integer(n),
          p = as.integer(p),
          Scatter = Scatter)$x
  z
}

File: pkg/src/whitening.c
/* $ID: whitening.c, last updated 2021-04-03, F.Osorio */

#include "fastmatrix.h"

void
whitening_chol(double *y, int *nrow, int *ncol, double *Scatter)
{ /* Cholesky whitening transformation  */
  char *side = "R", *uplo = "L", *trans = "T", *diag = "N";
  int info = 0, job = 0, n = *nrow, p = *ncol;

  FM_chol_decomp(Scatter, p, p, job, &info);
  if (info)
    error("DPOTRF in cholesky decomposition gave code %d", info);

  FM_invert_triangular(Scatter, p, p, job, &info);
  if (info)
    error("DTRTRI in matrix inversion gave code %d", info);

  BLAS3_trmm(1.0, Scatter, p, n, p, side, uplo, trans, diag, y, n);
}
